core/http-auth-aws-crt/src/main/java/software/amazon/awssdk/http/auth/aws/crt/internal/CrtHttpRequestConverter.java

```
@@ -0,0 +1,152 @@
+import software.amazon.awssdk.utils.StringUtils;
+import software.amazon.awssdk.utils.http.SdkHttpUtils;
+
+@SdkInternalApi
+public final class CrtHttpRequestConverter {
+
+    private static final int READ_BUFFER_SIZE = 4096;
+
+    private CrtHttpRequestConverter() {
+    }
+
+    /**
+     * Convert an {@link SdkHttpRequest} to an {@link HttpRequest}.
+     */
+    public static HttpRequest toRequest(SdkHttpRequest request, ContentStreamProvider payload) {
+        String method = request.method().name();
+        String encodedPath = encodedPathToCrtFormat(request.encodedPath());
+
+        String encodedQueryString = request.encodedQueryParameters().map(value -> "?" + value).orElse("");
+
+        HttpHeader[] crtHeaderArray = createHttpHeaderArray(request);
+
+        HttpRequestBodyStream crtInputStream = null;
+        if (payload != null) {
+            crtInputStream = new CrtInputStream(payload, READ_BUFFER_SIZE);
+        }
+
+        return new HttpRequest(method, encodedPath + encodedQueryString, crtHeaderArray, crtInputStream);
+    }
+
+    /**
+     * Convert an {@link HttpRequest} to an {@link SdkHttpRequest}.
+     */
+    public static SdkHttpRequest toRequest(SdkHttpRequest request, HttpRequest crtRequest) {
+        SdkHttpRequest.Builder builder = request.toBuilder();
+
+        builder.clearHeaders();
+        for (HttpHeader header : crtRequest.getHeaders()) {
+            builder.appendHeader(header.getName(), header.getValue());
+        }
+
+        URI fullUri;
+        try {
+            String portString = SdkHttpUtils.isUsingStandardPort(builder.protocol(), builder.port()) ? "" : ":" + builder.port();
+            String encodedPath = encodedPathFromCrtFormat(request.encodedPath(), crtRequest.getEncodedPath());
+            String fullUriString = builder.protocol() + "://" + builder.host() + portString + encodedPath;
+            fullUri = new URI(fullUriString);
+        } catch (URISyntaxException e) {
+            return null;
```

Can we throw here instead of returning `null`?. Rationale, while unlikely this might be if for whatever reason something changes that prevents the URI from being parsed correctly this will shadow that issue and instead move it to a NPE somewhere else.